### ISSUE TITLE
How to update reference with lookup ref

### ISSUE THREAD

**Issue by jethrokuan** [zcaudate/adi#15]
Suppose I have a schema like this:

```
(def schema
  {:user {:name [{:type :string}]
          :id   [{:type     :uuid
                  :required true
                  :unique   :identity}]}
   :book {:user [{:type     :ref
                  :required true
                  :ref      {:ns :user}}]}})
```

And I added a user, and have access to `id`, a unique attribute.

How would I construct an efficient query (lookup refs), to create a book? The query would look like this:

```
(adi/insert ds {:book {:name "Alice In Wonderland"
                       :user uuid}})
```

**Comment by zcaudate** [comment]
you will do something like this - you have to get the internal datomic id first:

    (let [id (adi/select {:user/id uuid} :return :ids :first)]
       (adi/insert ds {:book {:name "Alice In Wonderland"
                                           :user id}}))

**Comment by jethrokuan** [comment]
This will work for now, but something like:

```
(adi/insert ds {:book {:name "Alice In Wonderland" :user [:id uuid]}})
```

would be really nice, keeping in style with datomic's recommendation of lookup refs.

**Comment by zcaudate** [comment]
yeah i know... :/

it's hard to achieve... can you try if this works?

```
(adi/update! ds
             {:user/id uuid}
             {:user/books/name "Alice In Wonderland"})
```

**Comment by jethrokuan** [comment]
I'll give that a go at my earliest chance, and report back.

**Comment by jethrokuan** [comment]
I confirm that this works. However, this requires a different mental model, so I think i'd stick with the first recommendation.

**Comment by zcaudate** [comment]
great. the problem with the `insert` method is that it's ambiguous. we don't know if we are adding a new user or referencing an existing user (even if the id is unique, we can't know that before the insert). this is why I don't think it'll be changed to your suggested behavior.

**Comment by zcaudate** [comment]
actually... I noticed that you used a vector as opposed to a map... [:id uuid]

I've never seen that before. It's doable... but really really hard =p

**Comment by jethrokuan** [comment]
Perhaps this would provide some inspiration:

https://gist.github.com/robert-stuttaford/39d43c011e498542bcf8

**Comment by jethrokuan** [comment]
db ids are longs, while lookup refs are vectors. One could always assume that when it's a lookup ref, it's not an insert of a new user. Might look into this when I have more time.

**Comment by zcaudate** [comment]
is this locally supported by datomic? I haven't been looking at the newest stuff for a while now.

**Comment by jethrokuan** [comment]
Yes it is: here's an example of transaction data with lookup refs:

```
[{:db/id [:org/email "info@datomic.com"],
  :org/favoriteColor :blue}]
```

datomic requires that `:org/email` be a unique attribute.

**Comment by zcaudate** [comment]
okay... it should be easy to support then.
